### PR TITLE
fix(core): CloudFormation dynamic references can't be assigned to num…

### LINKED ISSUE #14824: (aws-quicksight): CfnDataSource.RedshiftParametersProperty will not accept dynamic references for "port"
When trying to create a Quicksight Redshift CfnDataSource resource, we have our Redshift connection information stored in SecretsManager, including endpoint, port, database name, username and password. When we try to use CDK's SecretsManager.secretValueFromJson(), it generates a CloudFormation {{resolve:secretsmanager}} dynamic reference. However, when CDK tries to synthesize the template, the CfnDataSource.DataSourceParameters.RedshiftParameters class requires port to be an Int, while the {{resolve}} string is a string, and it will fail synthesis with `port: "{{resolve:secretsmanager:arn:aws:secretsmanager:us-west-2:123456789012:secret:dummy-ABC123:SecretString:port::}}" should be a number.`

### Reproduction Steps

Create a Quicksight CfnDataSource and use SecretsManager to fill in the parameters:

```python
        quicksight.CfnDataSource(
            self,
            'redshiftDatasource-id',
            aws_account_id = core.Aws.ACCOUNT_ID,
            data_source_id = 'redshift',
            name = 'redshift-datasource',
            type = 'REDSHIFT',
            data_source_parameters=quicksight.CfnDataSource.DataSourceParametersProperty(
                redshift_parameters=quicksight.CfnDataSource.RedshiftParametersProperty(
                    host=redshift_secret.secret_value_from_json("endpoint").to_string(),
                    database=redshift_secret.secret_value_from_json("database").to_string(),
                    port=core.Token.as_number(redshift_secret.secret_value_from_json("database")),
                )
            ),
            credentials=quicksight.CfnDataSource.DataSourceCredentialsProperty(
                credential_pair=quicksight.CfnDataSource.CredentialPairProperty(
                    username=redshift_secret.secret_value_from_json("username").to_string(),
                    password=redshift_secret.secret_value_from_json("password").to_string()
                )
            )
        )
```

### What did you expect to happen?

The port field would accept a dynamic reference as well as a number.

### What actually happened?

Synthesis failed with `port: "{{resolve:secretsmanager:arn:aws:secretsmanager:us-west-2:123456789012:secret:dummy-ABC123:SecretString:port::}}" should be a number.`

### Environment

  - **CDK CLI Version  :** 1.103.0
  - **Framework Version:** 1.102.0
  - **Node.js Version:** 14.16.0
  - **OS               :** MacOS 11.2.3
  - **Language (Version):** Python 3.8.5

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by madeline-k**
…ber properties

We can't validate the type of dynamic references at synth time, since they will not be resolved until deploy time. This commit removes the type validation check for strings in the format: '{{resolve:...}'.

fixes #14824

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by madeline-k** [comment]
Looking for guidance on the best way to test this change. I validated manually with a cdk app the repros the customer issue in https://github.com/aws/aws-cdk/issues/14824 and https://github.com/aws/aws-cdk/issues/14876.  But I am not sure whether it makes more sense to test this with unit tests or integ tests, and in which module(s) those tests should be added.

**Comment by skinny85** [comment]
@madeline-k I would add unit tests for this in the SecretsManager module, verifying that I can pass `Token.asNumber(mySecret. secretValue)` to a property typed as `number` (I would probably use some L1 property whose type is `number` in the test).

**Comment by madeline-k** [comment]
@skinny85 Thanks for the guidance! I added a unit test as you described, but that unit test does not actually fail prior to the change in this CR. The error occurs during synth time, so I also added an integ test which does fail prior to the change here. Let me know what you think!

File: packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts [hunk 1]
```diff
@@ -687,6 +687,17 @@ test('fromSecretCompleteArn - grants', () => {
...
+test('fromSecretCompleteArn - can be assigned to a property with type number', () => {
+  const secretArn = 'arn:aws:secretsmanager:eu-west-1:111111111111:secret:MySecret-f3gDy9';
+  const secret = secretsmanager.Secret.fromSecretCompleteArn(stack, 'Secret', secretArn);
+  new lambda.Function(stack, 'MyFunction', {
+    code: lambda.Code.fromInline('foo'),
+    handler: 'bar',
+    runtime: lambda.Runtime.NODEJS,
+    memorySize: cdk.Token.asNumber(secret.secretValueFromJson('LambdaFunctionMemorySize')),
+  });
+});
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts:699]: You need to synthesize the Stack for the error to be seen 🙂. Adding any assertion, like `expect(stack).to(haveResourceLike('AWS::Lambda::Function')`, or using `SynthUtils.synthesize()`, will do the trick 🙂.

**Review by skinny85 — CHANGES_REQUESTED** [review]
> @skinny85 Thanks for the guidance! I added a unit test as you described, but that unit test does not actually fail prior to the change in this CR. The error occurs during synth time, so I also added an integ test which does fail prior to the change here. Let me know what you think!

Yes, because your unit test currently doesn't synthesize the Stack 😉. I explained it in my comment below.

I would get rid of the integ test, and rely on the unit test here.

File: packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts (hunk 1 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts:699]: Oh makes sense, of course! Will update :)

File: packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts [hunk 2]
```diff
@@ -687,6 +687,19 @@ test('fromSecretCompleteArn - grants', () => {
...
+  const secretArn = 'arn:aws:secretsmanager:eu-west-1:111111111111:secret:MySecret-f3gDy9';
+  const secret = secretsmanager.Secret.fromSecretCompleteArn(stack, 'Secret', secretArn);
+  new lambda.Function(stack, 'MyFunction', {
+    code: lambda.Code.fromInline('foo'),
+    handler: 'bar',
+    runtime: lambda.Runtime.NODEJS,
+    memorySize: cdk.Token.asNumber(secret.secretValueFromJson('LambdaFunctionMemorySize')),
+  });
+
+  SynthUtils.synthesize(stack);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts:700]: Let's write an `expect` assertion here - that checks `memorySize` is `'{{::resolve:secretsmanager...`.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Production code looks great - one small comment on the tests!

File: packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts (hunk 2 above)
Comment (madeline-k) [reply packages/@aws-cdk/aws-secretsmanager/test/secret.test.ts]: Updated!

**Review by skinny85 — APPROVED** [review]
Beatiful!

**Pull request merged**